task/download/download.go

```diff
@@ -44,40 +45,66 @@ type downloader struct {
 // whether it should be re-cloned.
 func getHashOfRepo(repo *task.Repository) string {
 	data := fmt.Sprintf("%s|%s|%s|%s", repo.Clone, repo.Ref, repo.Sha, repo.Download)
+	return getHash(data)
+}
+
+func getHash(s string) string {
 	hash := sha256.New()
-	hash.Write([]byte(data))
+	hash.Write([]byte(s))
 	return hex.EncodeToString(hash.Sum(nil))
 }
 
-func (d *downloader) Download(ctx context.Context, repo *task.Repository) (string, error) {
-	if repo == nil {
-		return "", errors.New("no repository provided to download")
+func (d *downloader) Download(ctx context.Context, repo *task.Repository, urls *task.ExecutableUrls) (string, error) {
+	if urls != nil {
+		return d.handleDownloadExecutable(ctx, urls)
+	} else if repo != nil {
+		return d.handleDownloadRepo(ctx, repo)
 	}
-	log := logger.FromContext(ctx)
+	return "", errors.New("no repository or executable urls provided to download")
+}
+
+func (d *downloader) handleDownloadExecutable(ctx context.Context, urls *task.ExecutableUrls) (string, error) {
+	osWithArch := fmt.Sprintf("%s/%s", runtime.GOOS, runtime.GOARCH)
+	url, ok := (*urls)[osWithArch]
+	if !ok {
+		return "", fmt.Errorf("os and architecture [%s] is not specified in task.yml file", osWithArch)
+	}
+
+	dest := filepath.Join(d.getBaseDownloadDir(), getHash(url))
+
+	if cacheHit := d.isCacheHit(ctx, dest); cacheHit {
+		// exit if the artifact destination already exists
+		return d.getDownloadPath(url, dest), nil
+	}
+
+	binpath, err := d.downloadFile(ctx, url, dest)
```

Comment:
after download, can we also write the ExecutableDownload struct into a file. Every time do append.

In this way, we will be able to know what had been downloaded at when. I believe this can help debugging, if it's not too much work

Reply:
Sure, added it (see `d.logExecutableDownload` in line 95. Thanks for the suggestion!